File: PostInorder.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Dec 17 09:04:59 2019

@author: tanvirkaur
"""

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
# Time complexity = O(n)
# Space complexity = O(n^2)
class Solution(object):
    def buildTree(self, inorder, postorder):
        """
        :type inorder: List[int]
        :type postorder: List[int]
        :rtype: TreeNode
        """
        if not inorder or not postorder:
            return None
        root = TreeNode(postorder[len(postorder)-1])
        indexx = inorder.index(root.val)
        root.left = self.buildTree(inorder[:indexx], postorder[:indexx])
        root.right = self.buildTree(inorder[indexx+1:], postorder[indexx:len(postorder)-1])
        return root

File: Post_inorder.java
//TC - O(n)
//SC - O(n)
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        HashMap<Integer,Integer> map = new HashMap<>();
        int idx = postorder.length-1;
        for(int  i = 0;i<inorder.length;i++){
            map.put(inorder[i],i);
        }
       return contructTree(inorder,postorder,0,inorder.length-1,map,idx);
    }
    
    public TreeNode contructTree(int[] inorder,int[] postorder,int start,int end, HashMap<Integer,Integer> map,int idx){
        
        if(start>end || idx<0) return null;
        
        TreeNode root = new TreeNode(postorder[idx]);
        int curr = map.get(root.val);
        
        root.left = contructTree(inorder,postorder,start,curr-1,map,idx-(end-curr+1));
        root.right = contructTree(inorder,postorder,curr+1,end,map,idx-1);
                
        return root;
    }
}